compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2204,16 +2202,24 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         case _ =>
       }
       // try an implicit conversion
-      inferView(tree, pt) match {
-        case SearchSuccess(inferred, _, _, _) =>
-          adapt(inferred, pt)(ctx.retractMode(Mode.ImplicitsEnabled))
-        case failure: SearchFailure =>
-          val prevConstraint = ctx.typerState.constraint
-          if (pt.isInstanceOf[ProtoType] && !failure.isInstanceOf[AmbiguousImplicits]) tree
-          else if (isFullyDefined(wtp, force = ForceDegree.all) &&
-                   ctx.typerState.constraint.ne(prevConstraint)) adapt(tree, pt)
-          else err.typeMismatch(tree, pt, failure)
-      }
+      val prevConstraint = ctx.typerState.constraint
+      def recover(failure: SearchFailure) =
```

Comment:
Fix: Previous logic allowed some implicit failures to be discarded if the expected type was a ProtoType.